src/graph.rs

```diff
@@ -71,21 +84,37 @@ impl Graph {
     // new returns a new graph instance from the given parameters.
     // The graph's edges are not generated yet, call fill_drg to compute the edges.
     pub fn new(size: usize, seed: [u8; 32], algo: DRGAlgo) -> Graph {
+        // FIXME: To avoid changing too much code at the moment the `GraphSpec`
+        //  is built here, but ideally the consumer should already provide it.
+        let spec = GraphSpec { seed, size, algo };
+        let mut rng = ChaCha20Rng::from_seed(spec.seed);
+
+        Self::new_from_rng(spec, &mut rng)
+    }
+
+    // FIXME: The RNG is not always necessary so this function is misleading.
+    fn new_from_rng(spec: GraphSpec, rng: &mut ChaCha20Rng) -> Graph {
         let mut g = Graph {
-            algo,
-            seed,
-            parents: Vec::with_capacity(size),
-            size,
+            spec,
+            parents: Vec::with_capacity(spec.size),
             children: vec![],
         };
-        match g.algo {
-            DRGAlgo::BucketSample => g.bucket_sample(),
-            DRGAlgo::MetaBucket(degree) => g.meta_bucket(degree),
+        match g.algo() {
+            DRGAlgo::BucketSample => g.bucket_sample(rng),
+            DRGAlgo::MetaBucket(degree) => g.meta_bucket(degree, rng),
             DRGAlgo::KConnector(k) => g.connect_neighbors(k),
         }
         g
     }
 
+    // FIXME: This function should replace the old `new` as much as possible.
+    pub fn many_from_spec(spec: GraphSpec, rng: &mut ChaCha20Rng, n_graph: usize) -> Vec<Graph> {
```

Comment:
This needs to be extended to save/load graphs, using the hash of the `GraphSpec` as index.

Reply:
Tbh I implemented this feature but the loading was not significantly faster than creating it from scratch already. Imo it can still be useful but not necessary required imminently (just to help prioritizing tasks)